templates/shift-scheduling/main.go

```diff
@@ -1,370 +1,225 @@
-				for j := 0; j < nShifts; j++ {
-					if value, ok := shifts.Domain(s, j).Value(); ok {
-						workerMap[value] = true
-					}
-				}
+	return output, nil
+}
 
-				return s.Apply(workerCount.Set(len(workerMap)))
+func format(
+	solverSolution mip.Solution,
+	x model.MultiMap[mip.Bool, assignment],
+	assignments []assignment,
+) output {
+	if !solverSolution.IsOptimal() && !solverSolution.IsSubOptimal() {
+		return output{}
+	}
+	nextShiftSolution := output{}
+	usedWorkers := make(map[int]struct{})
+
+	for _, assignment := range assignments {
+		if solverSolution.Value(x.Get(assignment)) >= 0.9 {
+			nextShiftSolution.AssignedShifts = append(nextShiftSolution.AssignedShifts, outputAssignment{
+				Start:    assignment.Start,
+				End:      assignment.End,
+				WorkerID: assignment.Worker.ID,
 			})
-	}).Validate(func(s store.Store) bool {
-		// The store is operationally valid if exactly one worker is assigned
-		// to a shift and no constraints are violated.
-		if !shifts.Singleton(s) {
-			return false
-		}
-
-		return !constraint.isProvenInfeasible(s)
-	}).Bound(func(s store.Store) store.Bounds {
-		// In order to make the search more efficient we define bounds.
-		// Given a partial schedule, `Bounds` returns a lower and an upper bound
-		// around the final schedule value once all workers have been assigned.
-		// As a rule of thumb: the better the bounds, the faster the search.
-
-		// The way we create new stores is that we never remove
-		// workers from a fixed shift. Thus a trivial bound is
-		// to assign a new worker to each unassigned shift.
-		// Likewise an upper bound on happiness is that each new worker
-		// is also happy.
-		wcUpper := workerCount.Get(s)
-		happinessUpper := happiness.Get(s)
-		for i := 0; i < nShifts; i++ {
-			d := shifts.Domain(s, i)
-			if _, ok := d.Value(); !ok {
-				wcUpper++
-				happinessUpper++
+			if _, ok := usedWorkers[assignment.Worker.ID]; !ok {
+				usedWorkers[assignment.Worker.ID] = struct{}{}
 			}
 		}
-
-		// For the lower bound we simply take the current worker count.
-		// As we know we need at least distinct 3 workers per day, we can
-		// take the maximum.
-		wcLower := workerCount.Get(s)
-		if wcLower < 3 {
-			wcLower = 3
-		}
-
-		// To find a lower bound on happiness we assume that any shift that has
-		// potential workers who have preferences cannot meet these preferences.
-		happinessLower := happiness.Get(s)
-		for i := 0; i < nShifts; i++ {
-			d := shifts.Domain(s, i)
-			increase := 1
-			if d.Len() <= 1 {
-				continue
-			}
-			for _, v := range d.Slice() {
-				if _, ok := prefShiftType[v]; ok {
-					// there is one worker who has a preference
-					increase = 0
-					break
-				}
+	}
+	nextShiftSolution.NumberAssignedWorkers = len(usedWorkers)
+
+	return nextShiftSolution
+}
+
+func newMIPModel(
+	input input,
+	potentialAssignments []assignment,
+	potentialAssignmentsPerWorker map[int][]*assignment,
+	demandCovering map[int][]assignment,
```

Comment:
Why is `potentialAssignmentsPerWorker` a `map[int][]*assignment` and `demandCovering` a `map[int][]assignment`. Can we harmonize that?